mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,964 @@
+
+  // Stop splitting when each parent has only one child left.
+  bool oneChild = false;
+  while (!oneChild) {
+    loopNest.clear();
+    forTree.clear();
+    forOperations.clear();
+    oneChild = true;
+
+    // Walk the function to create a tree of affine.for operations.
+    funcOp.walk([&](AffineForOp op) {
+      loopNest.push_back(op);
+      if (op.getParentOp()->getName().getStringRef() == "affine.for")
+        forTree[op.getOperation()->getParentOp()].push_back(op);
+
+      forOperations[op.getOperation()] = op;
+    });
+
+    // Separate one of the sibling at a time.
+    for (auto &loopNest : forTree) {
+      // This loop nest has no sibling problem.
+      // Check the next loop nest.
+      if (loopNest.second.size() < 2)
+        continue;
+      oneChild = false;
+      separateSiblingAffineForOps(forOperations[loopNest.first],
+                                  loopNest.second[loopNest.second.size() - 1],
+                                  loopNest.second);
+      // We need to walk the function again. Structure of loop nests within
+      // funcOp body has changed.
+      break;
+    }
+  }
+  return;
+}
+
+/// Populates `loadAndStoreOps` with a list of all affine.load/affine.store ops
+/// in the loop nest rooted at `rootForOp`.
+static void getAllLoadStores(AffineForOp rootForOp,
+                             SmallVector<Operation *, 8> &loadAndStoreOps) {
+  rootForOp.getOperation()->walk([&](Operation *op) {
+    if (isa<AffineLoadOp>(op) || isa<AffineStoreOp>(op)) {
+      MemRefType memref;
+      if (isa<AffineLoadOp>(op)) {
+        AffineLoadOp loadOp = dyn_cast<AffineLoadOp>(*op);
+        memref = loadOp.getMemRefType();
+      } else {
+        AffineStoreOp storeOp = dyn_cast<AffineStoreOp>(*op);
+        memref = storeOp.getMemRefType();
+      }
+      loadAndStoreOps.push_back(op);
+    }
+  });
+}
+
+/// Finds out element size of all memrefs in loop nest and populates (protected)
+/// `elementSize` property.
+void LoopInterchange::getElementsSize(AffineForOp &rootForOp) {
+  bool isLoadStore = false;
+  rootForOp.getOperation()->walk([&](Operation *op) {
+    isLoadStore = false;
+    MemRefType memref;
+    if (isa<AffineLoadOp>(op)) {
+      AffineLoadOp loadOp = dyn_cast<AffineLoadOp>(*op);
+      memref = loadOp.getMemRefType();
+      isLoadStore = true;
+    } else if (isa<AffineStoreOp>(op)) {
+      AffineStoreOp storeOp = dyn_cast<AffineStoreOp>(*op);
+      memref = storeOp.getMemRefType();
+      isLoadStore = true;
+    }
+    if (isLoadStore) {
+      auto elementType = memref.getElementType();
+      unsigned sizeInBits = 0;
+      if (elementType.isIntOrFloat())
+        sizeInBits = elementType.getIntOrFloatBitWidth();
+  
+      // Store element size to be later used to check whether two
+      // accesses are within a cache-line/elementSize distance apart
+      // for a useful locality.
+      this->elementSize[op] = llvm::divideCeil(sizeInBits, 8);
```

Comment:
There is already a function available to compute the size in bytes. See elsewhere.

Reply:
Now element size can be deduced directly from memrefType object. No need to get elementType object. However did not find any function to compute size in bytes - in MemRefType class and mlir::Types.h
Will look at other files.

Reply:
Done. Thanks.